File: kmp/iosApp/iosApp/FhirStore.swift
//
// Created by Arthur Nsereko Kahwa on 2026-05-09.
//

import Foundation
import Shared

@Observable
final class FhirStore {
    var patients: [Patient] = []
    var isLoading = false
    var error: String?

    private let client = FhirClient(baseUrl: "https://hapi.fhir.org/baseR4")

    func search(family: String) async {
        isLoading = true
        defer { isLoading = false }
        do {
            patients = try await client.searchPatients(family: family)
            error = nil
        } catch {
            self.error = error.localizedDescription
        }
    }
}

File: kmp/composeApp/src/androidMain/kotlin/com/example/fhirpatients/FhirStore.kt
package com.example.fhirpatients

import androidx.compose.runtime.getValue
import androidx.compose.runtime.mutableStateOf
import androidx.compose.runtime.setValue

class FhirStore(baseUrl: String = "https://hapi.fhir.org/baseR4") {

    private val client = FhirClient(baseUrl)

    var patients by mutableStateOf<List<Patient>>(emptyList())
        private set

    var isLoading by mutableStateOf(false)
        private set

    var error by mutableStateOf<String?>(null)
        private set

    suspend fun search(family: String) {
        isLoading = true
        try {
            patients = client.searchPatients(family)
            error = null
        } catch (t: Throwable) {
            error = t.message ?: "Unknown error"
        } finally {
            isLoading = false
        }
    }
}